Use Memcache's cas command to ensure consistency

brainix opened the pull request:
One of @reddit's use-cases is to allow multiple threads to update a Bloom filter.  Account for this by using [Memcache's `cas` (Check And Set) command](https://github.com/memcached/memcached/wiki/Commands#cas) to implement [optimistic concurrency control](https://en.wikipedia.org/wiki/Optimistic_concurrency_control) with a retry decorator.

As of this PR, one thread's changes to a `BloomFilter` will never stomp another thread's changes to the `BloomFilter`.  Instead, all `BloomFilter` updates will get merged together across threads.

brainix committed:
Use Memcache's cas command to ensure consistency

One of @reddit's use-cases is to allow multiple threads to update a
Bloom filter.  Account for this by using Memcache's cas (Check And Set)
command [1] to implement optimistic concurrency control [2] with a retry
decorator.

As of this PR, one thread's changes to a BloomFilter will never stomp
another thread's changes to the BloomFilter.  Instead, all BloomFilter
updates will get merged together across threads.

[1] cas: https://github.com/memcached/memcached/wiki/Commands#cas
[2] Optimistic concurrency control: https://en.wikipedia.org/wiki/Optimistic_concurrency_control

brainix committed:
In Python 2, prefer xrange() over range()

brainix committed:
Choose more descriptive identifier name

brainix committed:
Clean up unit test output

brainix committed:
Link to Medium post on how they use Bloom filters

dellis23 commented on bloom/exceptions.py:
```diff
@@ -0,0 +1,32 @@
+            self.key,
+            self.retriable,
+        )
+
```
Should be two blank lines here!

dellis23 replied:
:P

brainix replied:
[Good call.](https://www.python.org/dev/peps/pep-0008/#blank-lines)  [Addressed.](https://github.com/brainix/bloom-filter/pull/16/commits/c5b0495dc204fc88018ae90dca9fecb75d90a226)

dellis23 commented on bloom/bloom.py:
```diff
@@ -205,22 +207,41 @@ def _bit_offsets(self, value):
         else:
-            self.memcache.delete(self.key)
+            if not self.memcache.cas(self.key, bit_string, self._cas):
+                raise CheckAndSetError(memcache=self.memcache, key=self.key)
```
Will this only fail if it's a check and set error specifically?  I'm assuming something like a connection error would raise an exception rather than give a false value?  I think retrying instantly is fine in the case of lock contention, but I just don't want there to be a retry storm by every single client if an instance is getting overwhelmed or is unreachable.

brainix replied:
Excellent point!  I just did some testing to confirm the behavior of `cas()`:

    >>> from bloom import BloomFilter
    >>> recently_consumed = BloomFilter(key='recently-consumed')
    >>> # Kill the Memcached service...
    ...
    >>> recently_consumed.add('t3_1')
    Traceback (most recent call last):
      File "<stdin>", line 1, in <module>
      File "bloom/bloom.py", line 255, in add
        self.update({value})
      File "bloom/bloom.py", line 237, in wrapper
        return func(self, *args, **kwargs)
      File "bloom/bloom.py", line 272, in update
        self._store_bit_array()
      File "bloom/bloom.py", line 227, in _store_bit_array
        if not self.memcache.cas(self.key, bit_string, self._cas):
      File "/Users/raj/Documents/Code/reddit/bloom-filter/venv/lib/python2.7/site-packages/pymemcache/client/base.py", line 419, in cas
        return self._store_cmd(b'cas', key, expire, noreply, value, cas)
      File "/Users/raj/Documents/Code/reddit/bloom-filter/venv/lib/python2.7/site-packages/pymemcache/client/base.py", line 800, in _store_cmd
        buf, line = _readline(self.sock, buf)
      File "/Users/raj/Documents/Code/reddit/bloom-filter/venv/lib/python2.7/site-packages/pymemcache/client/base.py", line 1101, in _readline
        raise MemcacheUnexpectedCloseError()
    pymemcache.exceptions.MemcacheUnexpectedCloseError

The behavior above is also noted in the [pymemcache docs](http://pymemcache.readthedocs.io/en/latest/apidoc/pymemcache.client.base.html?highlight=cas#pymemcache.client.base.Client.cas) &mdash; the `cas()` command returns `True` on success (the `cas` value was the same, so the key was updated), `False` on failure (the `cas` value was different, so the key wasn&rsquo;t updated), or `None` if the key doesn&rsquo;t exist.  Any other error with network connectivity or with Memcached itself results in a [pymemcache exception](http://pymemcache.readthedocs.io/en/latest/apidoc/pymemcache.exceptions.html?highlight=exceptions).

So to answer your question, we only retry here in the case of lock contention.  Therefore, I don&rsquo;t think that we&rsquo;ll cause a retry storm.  However, if something goes wrong with the network or with Memcached, then we&rsquo;ll see exceptions in the listing service logs and in Sentry.  We may want to consider what to do in such cases to avoid polluting our logs and more importantly so that we don&rsquo;t lose any updates to recently consumed, but that&rsquo;s probably something that we can best strategize during a walk to Starbucks.

Does this seem reasonable, 💇 @dellis23?

dellis23 replied:
Yup, looks good!

brainix committed:
Update copyright year

brainix committed:
Conform to PEP 8's recommendation on blank lines

Surround top-level function and class definitions with two blank lines.

https://www.python.org/dev/peps/pep-0008/#blank-lines

brainix commented:
🐟
